internal/storage/databasestorage/databasestorage.go

```diff
@@ -0,0 +1,254 @@
+package databasestorage
+
+import (
+	"database/sql"
+	"errors"
+	"fmt"
+	"time"
+	"ya-prac-project1/internal/logger"
+	"ya-prac-project1/internal/metrics"
+
+	"github.com/jackc/pgerrcode"
+	pgx "github.com/jackc/pgx/v5/pgconn"
+	_ "github.com/jackc/pgx/v5/stdlib"
+	"go.uber.org/zap"
+)
+
+var database *sql.DB
+
+type Storage struct {
+}
+
+func NewStorage(dsn string) (*Storage, error) {
+
+	if err := initDB(dsn); err != nil {
+		return nil, err
+	}
+	prepareDB()
+
+	return &Storage{}, nil
+}
+
+func (s *Storage) SetValue(metricType, name, value string) error {
+	err := checkWrongType(metricType)
+	if err != nil {
+		return err
+	}
+
+	metric := s.GetMetric(metricType, name)
+	fmt.Println("metric", metric)
+	if metric == nil {
+		metric = &metrics.Metrics{
+			MType: metricType,
+			ID:    name,
+		}
+		s.AddMetric(metric)
+	}
+
+	metric.SetValue(value)
+	s.UpdateMetric(metric)
+	return nil
+}
+
+func (s *Storage) GetValue(metricType, name string) (string, error) {
+	err := checkWrongType(metricType)
+	if err != nil {
+		return "", err
+	}
+
+	metric := s.GetMetric(metricType, name)
+	if metric == nil {
+		return "", errors.New("not found metric")
+	}
+
+	return metric.GetValue(), nil
+}
+
+func (s *Storage) GetMetrics() []*metrics.Metrics {
```

Comment:
Так как сам слайс будет передавать по ссылке нет необходимости хранить в слайсе указатели нас структуры, так как копирования все равно не будет происходить, во всяком случае если это не требуется явно.
```suggestion
func (s *Storage) GetMetrics() []metrics.Metrics {
```

Reply:
Да, согласен, переделал.